File: unittests/libtests/topology/TestJacobian.cc
// -*- C++ -*-
//
// ----------------------------------------------------------------------
//
// Brad T. Aagaard, U.S. Geological Survey
// Charles A. Williams, GNS Science
// Matthew G. Knepley, University of Chicago
//
// This code was developed as part of the Computational Infrastructure
// for Geodynamics (http://geodynamics.org).
//
// Copyright (c) 2010-2015 University of California, Davis
//
// See COPYING for license information.
//
// ----------------------------------------------------------------------
//

#include <portinfo>

#include "TestJacobian.hh" // Implementation of class methods

#include "pylith/topology/Jacobian.hh" // USES Jacobian

#include "pylith/topology/Mesh.hh" // USES Mesh
#include "pylith/topology/Field.hh" // USES Field
#include "pylith/topology/SolutionFields.hh" // USES SolutionFields

#include "pylith/meshio/MeshIOAscii.hh" // USES MeshIOAscii

// ----------------------------------------------------------------------
CPPUNIT_TEST_SUITE_REGISTRATION( pylith::topology::TestJacobian );

// ----------------------------------------------------------------------
// Test constructor.
void
pylith::topology::TestJacobian::testConstructor(void)
{ // testConstructor
  PYLITH_METHOD_BEGIN;

  Mesh mesh;
  _initializeMesh(&mesh);
  Field field(mesh);
  _initializeField(&mesh, &field);
  Jacobian jacobian(field);

  Jacobian jacobianB(field, "baij");
  Jacobian jacobianC(field, "baij", true);
  Jacobian jacobianD(field, "sbaij");
  Jacobian jacobianE(field, "sbaij", true);

  PYLITH_METHOD_END;
} // testConstructor

// ----------------------------------------------------------------------
// Test constructor with subdomain field
void
pylith::topology::TestJacobian::testConstructorSubDomain(void)
{ // testConstructorSubDomain
  PYLITH_METHOD_BEGIN;

  Mesh mesh;
  _initializeMesh(&mesh);
  Field field(mesh);
  _initializeField(&mesh, &field);

  Mesh submesh(mesh, "bc");
  Field subfield(submesh);
  subfield.newSection(FieldBase::VERTICES_FIELD, submesh.dimension());
  subfield.allocate();
  subfield.zero();

  Jacobian jacobian(subfield);

  PYLITH_METHOD_END;
} // testConstructorSubDomain

// ----------------------------------------------------------------------
// Test matrix().
void
pylith::topology::TestJacobian::testMatrix(void)
{ // testMatrix
  PYLITH_METHOD_BEGIN;

  Mesh mesh;
  _initializeMesh(&mesh);
  Field field(mesh);
  _initializeField(&mesh, &field);
  Jacobian jacobian(field);

  const PetscMat matrix = jacobian.matrix();
  CPPUNIT_ASSERT(matrix);

  PYLITH_METHOD_END;
} // testMatrix

// ----------------------------------------------------------------------
// Test assemble().
void
pylith::topology::TestJacobian::testAssemble(void)
{ // testAssemble
  PYLITH_METHOD_BEGIN;

  Mesh mesh;
  _initializeMesh(&mesh);
  Field field(mesh);
  _initializeField(&mesh, &field);
  Jacobian jacobian(field);

  jacobian.assemble("flush_assembly");
  jacobian.assemble("final_assembly");

  PYLITH_METHOD_END;
} // testAssemble

// ----------------------------------------------------------------------
// Test zero().
void
pylith::topology::TestJacobian::testZero(void)
{ // testZero
  PYLITH_METHOD_BEGIN;

  Mesh mesh;
  _initializeMesh(&mesh);
  Field field(mesh);
  _initializeField(&mesh, &field);
  Jacobian jacobian(field);

  jacobian.zero();

  PYLITH_METHOD_END;
} // testZero

// ----------------------------------------------------------------------
// Test view().
void
pylith::topology::TestJacobian::testView(void)
{ // testView
  PYLITH_METHOD_BEGIN;

  Mesh mesh;
  _initializeMesh(&mesh);
  Field field(mesh);
  _initializeField(&mesh, &field);
  Jacobian jacobian(field);

  jacobian.assemble("final_assembly");

  jacobian.view();

  PYLITH_METHOD_END;
} // testView

// ----------------------------------------------------------------------
// Test write().
void
pylith::topology::TestJacobian::testWrite(void)
{ // testWrite
  PYLITH_METHOD_BEGIN;

  Mesh mesh;
  _initializeMesh(&mesh);
  Field field(mesh);
  _initializeField(&mesh, &field);
  Jacobian jacobian(field);

  jacobian.assemble("final_assembly");

  jacobian.write("jacobian.mat", mesh.comm());

  PYLITH_METHOD_END;
} // testWrite

// ----------------------------------------------------------------------
void
pylith::topology::TestJacobian::_initializeMesh(Mesh* mesh) const
{ // _initializeMesh
  PYLITH_METHOD_BEGIN;

  CPPUNIT_ASSERT(mesh);

  meshio::MeshIOAscii iohandler;
  iohandler.filename("data/tri3.mesh");
  iohandler.read(mesh);

  PYLITH_METHOD_END;
} // _initializeMesh

// ----------------------------------------------------------------------
void
pylith::topology::TestJacobian::_initializeField(Mesh* mesh,
                                                 Field* field) const
{ // _initializeField
  PYLITH_METHOD_BEGIN;

  CPPUNIT_ASSERT(mesh);
  CPPUNIT_ASSERT(field);

  field->newSection(FieldBase::VERTICES_FIELD, mesh->dimension());
  field->allocate();
  field->zero();

  PYLITH_METHOD_END;
} // _initializeField

// End of file 

File: unittests/pytests/topology/TestJacobian.py
#!/usr/bin/env python
#
# ======================================================================
#
# Brad T. Aagaard, U.S. Geological Survey
# Charles A. Williams, GNS Science
# Matthew G. Knepley, University of Chicago
#
# This code was developed as part of the Computational Infrastructure
# for Geodynamics (http://geodynamics.org).
#
# Copyright (c) 2010-2015 University of California, Davis
#
# See COPYING for license information.
#
# ======================================================================
#

## @file unittests/pytests/topology/TestJacobian.py

## @brief Unit testing of Jacobian object.

import unittest

from pylith.topology.Jacobian import Jacobian


# ----------------------------------------------------------------------
class TestJacobian(unittest.TestCase):
  """
  Unit testing of Jacobian object.
  """

  def setUp(self):
    """
    Setup mesh and associated field.
    """
    from spatialdata.geocoords.CSCart import CSCart
    cs = CSCart()
    cs.inventory.spaceDim = 2
    cs._configure()

    from spatialdata.units.Nondimensional import Nondimensional
    normalizer = Nondimensional()
    normalizer._configure()    

    from pylith.meshio.MeshIOAscii import MeshIOAscii
    importer = MeshIOAscii()
    importer.inventory.filename = "data/tri3.mesh"
    importer.inventory.coordsys = cs
    importer._configure()
    self.mesh = importer.read(debug=False, interpolate=False)

    from pylith.topology.SolutionFields import SolutionFields
    fields = SolutionFields(self.mesh)
    fields.add("disp t+dt", "displacement")
    fields.solutionName("disp t+dt")
    solution = fields.solution()
    solution.newSection(solution.VERTICES_FIELD, self.mesh.dimension())
    solution.allocate()
    solution.zero()

    self.fields = fields
    self.jacobian = Jacobian(solution)
    return


  def tearDown(self):
    self.jacobian.cleanup()
    return


  def test_constructor(self):
    """
    Test constructor.
    """
    # setUp() tests constructor with default type
    jacobianA = Jacobian(self.fields.solution(), "aij")
    jacobianB = Jacobian(self.fields.solution(), "baij")
    return


  def test_matrix(self):
    """
    Test matrix().

    :WARNING: This is not a complete test of matrix(). We do not
    verify the results.
    """
    matrix = self.jacobian.matrix()

    # No testing of result.
    return


  def test_assemble(self):
    """
    Test assemble().

    :WARNING: This is not a complete test of assemble(). We do not
    verify the results.
    """
    self.jacobian.assemble("flush_assembly")
    self.jacobian.assemble("final_assembly")

    # No testing of result.
    return


  def test_zero(self):
    """
    Test zero().

    :WARNING: This is not a complete test of zero(). We do not
    verify the results.
    """
    self.jacobian.zero()

    # No testing of result.
    return


  def test_view(self):
    """
    Test view().

    :WARNING: This is not a complete test of view(). We do not
    verify the results.
    """
    self.jacobian.assemble("final_assembly")
    self.jacobian.view()

    # No testing of result.
    return


  def test_write(self):
    """
    Test write().

    :WARNING: This is not a complete test of write(). We do not
    verify the results.
    """
    self.jacobian = Jacobian(self.fields.solution(), "aij")
    self.jacobian.assemble("final_assembly")

    self.jacobian.write("jacobian.mat", self.mesh.comm())

    # No testing of result.
    return


# End of file 
